Triangle Streaming Server/ReceiveStream.cs

```diff
@@ -77,12 +72,39 @@ protected override void OnMessage(MessageEventArgs e)
 
 				Clients.Add(this.ID, streamToWatch);
 			}
-			else if(e.Data.Equals("LIST"))
+			else if(e.Data.StartsWith("CHALLENGE: "))
 			{
-				//Send a list of streams to the client.
-				var streams = JsonConvert.SerializeObject(StreamQueueManager.GetInstance().Streams.Values.ToList());
+				string challengeString = e.Data.Replace("CHALLENGE: ", "");
+
+				//Convert base64 string back to encrypted bytes
+				var challengeBytes = Convert.FromBase64String(challengeString);
 
-				this.Send(streams.ToString());
+				//Decrypt bytes into string
+				IAsymmetricBlockCipher eng = new Pkcs1Encoding(new RsaEngine());
+				eng.Init(false, _keyPair.Private);
+				var responseString = Encoding.UTF8.GetString(eng.ProcessBlock(challengeBytes, 0, challengeBytes.Length));
+
+				//Send original message back to client
+				this.Send(String.Format("CHALLENGERESPONSE: {0}", responseString));
+			}
+			else
+			{
+				switch(e.Data)
+				{
+					case "PUBKEY":
+						//Send public key to client
+						Send(_pubKey);
+						break;
+					case "LIST":
+						//Send a list of streams to the client.
+						var streams = JsonConvert.SerializeObject(StreamQueueManager.GetInstance().Streams.Values.ToList());
```

Comment:
Null reference should not be possible, the Streams dictionary is always initialized on program start. If there are no streams, an empty array is returned.